plugins/autoid/plugin.js

```diff
@@ -158,6 +160,209 @@
+          linkTypeTab.onChange = modifiedLinkTypeChanged;
+
+          // Modify main dialog 'onOk' function to handle heading data
+          def.onOk = modifiedOnOk;
+        }
+      }
+
+      function modifiedLinkTypeChanged() {
+        var dialog = this.getDialog(),
+          partIds = [ 'urlOptions', 'anchorOptions', 'emailOptions', 'headingOptions' ],
+          typeValue = this.getValue(),
+          uploadTab = dialog.definition.getContents( 'upload' ),
+          uploadInitiallyHidden = uploadTab && uploadTab.hidden;
+
+        if ( typeValue == 'url' ) {
+          if ( editor.config.linkShowTargetTab )
+            dialog.showPage( 'target' );
+          if ( !uploadInitiallyHidden )
+            dialog.showPage( 'upload' );
+          } else {
+            dialog.hidePage( 'target' );
+          if ( !uploadInitiallyHidden )
+            dialog.hidePage( 'upload' );
+        }
+
+        for ( var i = 0; i < partIds.length; i++ ) {
+          var element = dialog.getContentElement( 'info', partIds[ i ] );
+          if ( !element )
+            continue;
+
+          element = element.getElement().getParent().getParent();
+          if ( partIds[ i ] == typeValue + 'Options' )
+            element.show();
+          else
+            element.hide();
+        }
+
+        dialog.layout();
+      }
+
+      function createHeadingLinkContent() {
+        var headings;
+
+        return {
+          type: 'vbox',
+          id: 'headingOptions',
+          children: [ {
+            type: 'select',
+            id: 'heading',
+            label: 'Select a Heading',
+            style: 'width : 100%;',
+            items: [ [ '' ] ],
+            setup: function() {
+              this.clear();
+              this.add( [ '' ] );
+
+              headings = findAllHeadings();
+
+              if (headings && headings.count()) {
+                this.getElement().show();
+
+                for (var i = 0; i < headings.count(); i++) {
+                  var heading = headings.getItem(i);
+                  this.add( heading.getText(), heading.getId() );
+                }
+              }
+              else {
+                this.getElement().hide();
+              }
+            },
+            commit: function(data) {
+              if (!data.heading)
+                data.heading = {};
+
+              data.heading.id = this.getValue();
+              // set the data type to 'anchor' since the procedure is the same
+              // for headings.  This way, we don't have to modify the main
+              // 'plugin.js' for 'link' to handle a new link type.
+              if (data.heading.id)
+                data.type = 'anchor';
+            }
+          },
+          {
+            type: 'html',
+            id: 'noHeadings',
+            style: 'text-align: center;',
+            html: '<div role="note" tabIndex="-1">' + CKEDITOR.tools.htmlEncode( '(No headings available in the document)' ) + '</div>',
```

Comment:
Ah.  We've never really found a lot of their messages to be very user friendly...:/